src/persist/examples/kafka_upsert.rs

```diff
@@ -0,0 +1,562 @@
+    ),
+    Box<dyn Error>,
+>
+where
+    G: Scope<Timestamp = u64>,
+    S: Source + 'static,
+{
+    // Fetch start offsets.
+    // - TODO: Don't use read_to_end_flattened
+    let starting_offsets: Vec<_> = bindings_read
+        .snapshot()?
+        .read_to_end_flattened()?
+        .into_iter()
+        // TODO: look extra hard at whether we need < or <= here. According
+        // to how timely frontiers work we need <.
+        .filter(|(_binding, ts, _diff)| *ts < start_ts)
+        .map(|((source_timestamp, _assigned_timestamp), _ts, _diff)| source_timestamp)
+        .collect();
+
+    let mut source = S::new(starting_offsets, emission_interval_ms, now_fn);
+
+    let mut source_op = OperatorBuilder::new("read_source".to_string(), scope.clone());
+    let activator = scope.activator_for(&source_op.operator_info().address[..]);
+
+    let (mut records_out, records) = source_op.new_output();
+    let (mut bindings_out, bindings) = source_op.new_output();
+
+    let mut current_ts = now_fn();
+
+    source_op.build(move |mut capabilities| {
+        let mut records_capability = capabilities.remove(0);
+        let mut bindings_capability = capabilities.remove(0);
+
+        move |_frontiers| {
+            // only emit a binding update if we actually have a current offset
+            let now = now_fn();
+            let now_clamped = now - (now % timestamp_interval_ms);
+            if now_clamped != current_ts {
+                // emit bindings for all offsets we're currently aware of
+                let mut bindings_out = bindings_out.activate();
+                let mut bindings_session = bindings_out.session(&bindings_capability);
+
+                for source_timestamp in source.frontier() {
+                    bindings_session.give((
+                        (source_timestamp, AssignedTimestamp(current_ts)),
+                        current_ts,
+                        1,
+                    ));
+                }
+
+                // then update to new binding
+                current_ts = now_clamped;
+                bindings_capability.downgrade(&current_ts);
+                records_capability.downgrade(&current_ts);
+            }
+
+            if let Some((record, _source_timestamp)) = source.get_next_message() {
+                let mut records_out = records_out.activate();
+                let mut records_session = records_out.session(&records_capability);
+                records_session.give((record, current_ts, 1));
+            }
+
+            activator.activate_after(Duration::from_millis(100));
+        }
+    });
+
+    Ok((records, bindings))
+}
+
+/// Persists bindings to the given handle and emits errors as a `Stream`.
+///
+/// Returns a stream of errors.
+// TODO: also return a Stream<G, ()> that could be used to listen on
+// the frontier. As an alternative to listening for `Seal` events on a stream handle.
+fn persist_bindings<G, ST>(
```

Comment:
It's true about upsert, I have an ongoing branch where I factored the upsert operator out of `upsert.rs` and moved it to the `persist` crate. I would change the example to use that once it's merged.

About the `Persist` operator: yes, It's mostly true but `persist_bindings` is more like a sink in that it doesn't forward the persisted data. I could add that to our operator library as well, though.

Reply:
WDYT, should I turn the current `Persist` operator into a `PersistentSink` (`PersistenceSink`? `PersistSink`)?

Reply:
Ack on the upsert operator, doing that later sounds fine to me!

At some point, we'll want to allow mz to use some persisted collection of another instance as a SOURCE (for arjun's chaining of mz instances). This suggests that we might want an mz-level SINK that persists its input and as a result, we might want a PersistedSink operator (naming parallel to PersistedSource). So that seems like a fine generalization to me for this.